### DISCUSSION TITLE
Execute Job, No Tenant is not resolved.

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#1058] [Q&A] [answered]
I am executing this code in a scheduled job:

        public override string ExecuteJob(IServiceProvider provider)
        {
            string log = string.Empty;
            log += "ExperimentalJob: start";

            //From Repository Oqtane
            try
            {
                log += "ExperimentalJob: GetRequiredService<IAliasRepository>()";
                var aliasRepository = provider.GetRequiredService<IAliasRepository>();
                List<Alias> aliases = aliasRepository.GetAliases().ToList();
                log += "<br /> Experimentals Returned: " + aliases.Count().ToString();
            }
            catch (Exception ex)
            {
                log += "<br /> Experimentals Error: " + ex.Message + "<br />" ;
            }

            //From Repostitory Module
            try
            {
                log += "ExperimentalJob: GetRequiredService<IExperimentalRepository>()";
                var experimentalRepostitory = provider.GetRequiredService<IExperimentalRepository>();
                List<Models.Experimental> experimentallist = experimentalRepostitory.GetExperimentals(28).ToList();
                log += "<br /> Experimentals Returned: " + experimentallist.Count().ToString();
            }
            catch (Exception ex)
            {
                log += "<br /> Experimentals Error: " + ex.Message + "<br />";
            }

            log += "<br /> ExperimentalJob: stop";

            return log;
        }

The first try block executes just fine the second block however does not.  I have tracked it down to the Tenant resolver.  In the second try block the tenant is not resolved and errors with object not set error.

The code is picked from the DNF.Projects module.  One difference between DNF and Experimental modules is the Entity Framwork versions.

**Comment by sbwalker** [comment, marked as answer]
If you look at the code in the DNF.Projects GithubJob - or the NotificationJob in the core framework you will notice the following lines of code:

```
                // use the SiteState to set the Alias explicitly so the tenant can be resolved
                var siteState = provider.GetRequiredService<SiteState>();
                siteState.Alias = alias;
```

Like the comment indicates, this logic is required in order to set the Alias so that the tenant can be resolved. This is necessary because a scheduled job is not invoked by a client request and therefore is not tenant aware. Your scheduled job does not contain this logic, so it will not work.

**Reply by leigh-pointer** [reply]
Thank you, everything is good now.

**Reply by sbwalker** [reply]
@leigh-pointer you may like the improvement in #1061 as it simplifies job development

**Reply by leigh-pointer** [reply]
Yeah! im going look at this tomorrow! Sounds a lot more intuitive that the current implementation.

Thanks!

**Reply by leigh-pointer** [reply]
@sbwalker #1061 makes life a lot easier for job development.
